src/OpenTelemetry.Exporter.Prometheus.HttpListener/Internal/PrometheusCollectionManager.cs

```diff
@@ -170,6 +172,39 @@ private ExportResult OnCollect(Batch<Metric> metrics)
 
         try
         {
+            this.scopes.Clear();
+
+            foreach (var metric in metrics)
+            {
+                if (PrometheusSerializer.CanWriteMetric(metric))
+                {
+                    this.scopes.Add(metric.MeterName);
+                }
+            }
+
+            foreach (var scope in this.scopes)
```

Comment:
Instead of waiting to populate the HashSet first, and then writing the scope info, you could write the scope info as you're adding elements to the HashSet by checking the result of `Hashet.Add`.

```csharp
foreach (var metric in metrics)
{
    if (PrometheusSerializer.CanWriteMetric(metric))
    {
        if (this.scopes.Add(metric.MeterName))
        {
            try
            {
                cursor = PrometheusSerializer.WriteScopeInfo(this.buffer, cursor, scope);
                
                break;
            }
            catch (IndexOutOfRangeException)
            {
                if (!this.IncreaseBufferSize())
                {
                    // there are two cases we might run into the following condition:
                    // 1. we have many metrics to be exported - in this case we probably want
                    //    to put some upper limit and allow the user to configure it.
                    // 2. we got an IndexOutOfRangeException which was triggered by some other
                    //    code instead of the buffer[cursor++] - in this case we should give up
                    //    at certain point rather than allocating like crazy.
                    throw;
                }
            }
        }
    }
}
```